docs/design/repair-with-hashes.md

```diff
@@ -0,0 +1,63 @@
+# Repair using Piece Hashes
+
+## Abstract
+
+The satellite should repair files also using piece hashes to minimize CPU and bandwidth load.
+
+## Background
+
+The white-paper states:
+
+> Data repair is an ongoing, costly operation that will use significant bandwidth, memory, and processing power, often impacting a single operator. As a result, repair resource usage should be aggressively minimized as much as possible.
+> For repairing a segment to be effective at minimizing bandwidth usage, only as few pieces as needed for reconstruction should be downloaded. Unfortunately, Reed-Solomon is insufficient on its own for correcting errors when only a few redundant pieces are provided. Instead, piece hashes provide a better way to be confident that we’re repairing the data correctly.
+> To solve this problem, hashes of every piece will be stored alongside each piece on each storage node. A validation hash that the set of hashes is correct will be stored in the pointer. During repair, the hashes of every piece can be retrieved and validated for correctness against the pointer, thus allowing each piece to be validated in its entirety. This allows the repair system to correctly assess whether or not repair has been completed successfully without using extra redundancy for the same task.
+
+Hash verification on the satellite requires understanding the current piece signing and verification workflow:
+
+1. Satellite generates a new key-pair for each piece.
+2. The piece public key is included in the order limit and signed by the satellite
+3. The piece private key is included in the create segment, download segment or delete a segment.
+4. The uplink uses piece private key to sign the orders.
```

Comment:
If the key pair is generated on the satellite, why is it the uplink that does the signing here? Maybe I am just misunderstanding.

Reply:
piece private key is used by uplink to sign piece hash which contains hash of uploaded data and order (not order limit) which informs how much bandwidth was used (put/get). Sending such data to satellite to sign make no sense.

Reply:
Step #1 says "Satellite generates a new key-pair for each piece". Why does the satellite generate the pair if the uplink is the one using it?

Reply:
The satellite generates the key pair so that the public key can be signed by the satellite, so that the satellite can trust the signing done with it.

I confess this text is stolen from https://github.com/storj/storj/pull/2525; maybe looking at that PR would help.

Reply:
Thanks for the link. That clarified things.